### ISSUE TITLE
OptionalUnit conflicts with explicitApi=strict

### ISSUE THREAD

**Issue by pettermahlen** [detekt/detekt#4647]
## Expected Behavior

It should be possible to enable [strict explicit API mode](https://kotlinlang.org/docs/whatsnew14.html#explicit-api-mode-for-library-authors) without getting conflicts with Detekt rules.

## Observed Behavior

The OptionalUnit rule, if activated, considers an explicitly declared Unit return type as an error, and the compilation fails if the return type is implicit.

## Steps to Reproduce

Enable both explicitApi = "strict" and OptionalUnit.

**Comment by marschwar** [comment]
Thank you for bringing this up. I believe this rule is much older than the `explicitApi` mode and as such unaware of this setting. Personally I do not see many options here other than not enabling the rule as it will - by definition - cause your build to fail. 

I do not think it is a bad rule but it may not make sense for every project. I guess that is why it is not enabled by default.

Do you have any other suggestions? Please let us know.

**Comment by pettermahlen** [comment]
We have a large repository, 1M+ LOC, where we want to have consistent style checks (so a single set of Detekt rules), but where it doesn't make sense to enable strict API everywhere. So ideally, the OptionalUnit rule would be aware of explicitApi setting, disabling itself when it is not 'off'. I'm not sure if that's even possible?

**Comment by cortinico** [comment]
Hi @pettermahlen 👋 

> So ideally, the OptionalUnit rule would be aware of explicitApi setting, disabling itself when it is not 'off'. I'm not sure if that's even possible?

It's possible. The problem is that you'll have to run Detekt with type resolution, and I don't know if this is the case for you (@davidburstromspotify can maybe say?).

Moreover, we're also blocked on this side because of this:
- https://github.com/detekt/detekt/issues/4465

The tl;dr is that the compilation flag you specify once you enable `explicitApi` doesn't get propagate properly to Detekt, so we're effectively compiling your code without that flag. 

More on this also here:
- https://github.com/detekt/detekt/issues/3874

So yeah, it's doable, but it will take a couple of releases for this to even get implemented and reach the public.

For the timebeing, you should resort to locally suppressing inspections, or defining an `exclude` in the config that will exclude an entire folder (which I believe is where your library lives or so).

**Comment by pettermahlen** [comment]
Thanks for the response, for the time being, we've disabled the `OptionalUnit` rule, so no problem for us. :) It's not the most important rule and off by default, so it seemed like an easy way out.

**Comment by davidburstromspotify** [comment]
Hi! No, we're not running Detekt with type resolution at this point.

**Comment by cortinico** [comment]
> Hi! No, we're not running Detekt with type resolution at this point.

Gotcha. So this rules out any access to the compiler flags

**Comment by cortinico** [comment]
I think we should close this as this is essentially a transient issue. Will be fixed by https://github.com/detekt/detekt/issues/4465

Moreover, this capability will rely on having Type Resolution enabled.

**Comment by 3flex** [comment]
I'm reopening as I'm working through PRs to address #4465. The PRs are only about setting the compiler flags. Some rules (like OptionalUnit) will need updating with new logic to take advantage of the correct values being passed through, which I'm not addressing in those PRs.

The PR to set & pass through explicit API is not done yet but will be soon, I'll link it here once open.

**Comment by 3flex** [comment]
Support for this is being added in https://github.com/detekt/detekt/pull/7418, but the rule needs to be updated to take advantage of this.

Use one of the following:
```kotlin
compilerResources?.languageVersionSettings?.explicitApiEnabled
compilerResources?.languageVersionSettings?.getFlag(AnalysisFlags.explicitApiMode)
```

The first is suitable if the rule only cares if explicit API mode is in strict mode or not. The second can be used to get the specific value that's been set (strict, warning or disabled).

**Comment by cortinico** [comment]
> Support for this is being added in #7418, but the rule needs to be updated to take advantage of this.

@3flex can we close this as #7418 is merged?

**Comment by 3flex** [comment]
No, #7430 has the fix.
